File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_date_range/DateRangeSelection.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogSelection;

import java.time.OffsetDateTime;
import java.util.Arrays;

class DateRangeSelection extends BaseDialogSelection<OffsetDateTime[]> {
    public DateRangeSelection(OffsetDateTime[] current, OffsetDateTime[] draft) {
        super(current, draft);
        if (current == null)
            setCurrentSelection(new OffsetDateTime[]{null, null}, false);
        if (draft == null)
            setDraftSelection(new OffsetDateTime[]{null, null}, false);
    }

    @Override
    public boolean compareValues(OffsetDateTime[] v1, OffsetDateTime[] v2) {
        return Arrays.equals(v1, v2);
    }

    @Override
    public void setCurrentSelection(OffsetDateTime[] newValue, boolean notify) {
        if (newValue == null)
            newValue = new OffsetDateTime[]{null, null};
        super.setCurrentSelection(newValue, notify);
    }

    @Override
    public void setDraftSelection(OffsetDateTime[] newValue, boolean notify) {
        if (newValue == null)
            newValue = new OffsetDateTime[]{null, null};
        super.setDraftSelection(newValue, notify);
    }

    @Override
    public boolean hasCurrentSelection() {
        OffsetDateTime[] current = getCurrentSelection();
        return current != null && current[0] != null && current[1] != null;
    }

    @Override
    public boolean hasDraftSelection() {
        OffsetDateTime[] draft = getDraftSelection();
        return draft != null && draft[0] != null && draft[1] != null;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_date_range/DateRangeSelection.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range

import android.os.Bundle
import com.github.rooneyandshadows.lightbulb.commons.utils.BundleUtils
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogSelection
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range.DateRangePickerDialog.DateRange

internal class DateRangeSelection(current: DateRange?, draft: DateRange?) :
    BaseDialogSelection<DateRange>(current, draft) {

    init {
        if (current == null) setCurrentSelection(null, false)
        if (draft == null) setDraftSelection(null, false)
    }

    @Override
    override fun compareValues(v1: DateRange?, v2: DateRange?): Boolean {
        if ((v1 == null) && (v2 == null)) return true
        if ((v1 == null) || (v2 == null)) return false
        return v1.compare(v2)
    }

    @Override
    override fun hasCurrentSelection(): Boolean {
        val current = getCurrentSelection()
        return current != null
    }

    @Override
    override fun hasDraftSelection(): Boolean {
        val draft = getDraftSelection()
        return draft != null
    }

    @Override
    override fun saveState(): Bundle {
        return Bundle().apply bundle@{
            getCurrentSelection()?.apply {
                BundleUtils.putParcelable(CURRENT_SELECTION_STATE_KEY, this@bundle, this)
            }
            getDraftSelection()?.apply {
                BundleUtils.putParcelable(DRAFT_SELECTION_STATE_KEY, this@bundle, this)
            }
        }
    }

    @Override
    override fun restoreState(source: Bundle) {
        val clazz = DateRange::class.java
        source.apply {
            BundleUtils.getParcelable(CURRENT_SELECTION_STATE_KEY, this, clazz)?.apply {
                setCurrentSelection(this, false)
            }
            BundleUtils.getParcelable(DRAFT_SELECTION_STATE_KEY, this, clazz)?.apply {
                setDraftSelection(this, false)
            }
        }
    }
}